File: leetcode/201-300/300-longest-increasing-subsequence/bench/lisscan.c
/* Benchmark mirror of lisscan.kara — LeetCode #300, Longest Increasing
 * Subsequence. Same patience sorting, same hand-written binary search, same
 * reused stack tails buffer. See ../README.md § Benchmarks. */
#include <stdio.h>
#include <stdlib.h>

#define N_ARRAYS 3000
#define LEN      512
#define PASSES   24
#define SPREAD   4096

static long lcg(long state) {
    return (state * 1103515245 + 12345) & 0x7fffffff;
}

int main(void) {
    long total = (long)N_ARRAYS * LEN;
    long *data = malloc(sizeof(long) * total);
    if (!data) return 1;

    long state = 20300;
    for (long i = 0; i < total; i++) {
        state = lcg(state);
        data[i] = (state / 65536) % SPREAD;
    }

    long tails[LEN];
    long checksum = 0;

    for (int pass = 0; pass < PASSES; pass++) {
        for (long a = 0; a < N_ARRAYS; a++) {
            long base = a * LEN;
            long n_tails = 0;

            for (long k = 0; k < LEN; k++) {
                long x = data[base + k];

                /* size_t, not long: the idiomatic C index type, and the fair one.
               With signed `long` here clang emits the sign-correction sequence
               for `/ 2` and the lane costs 1.249 s instead of 950 ms — a 1.31x
               handicap this mirror should not be carrying. See ../README.md. */
            size_t lo = 0, hi = (size_t)n_tails;
                while (lo < hi) {
                    size_t mid = lo + (hi - lo) / 2;
                    if (tails[mid] < x) lo = mid + 1;
                    else hi = mid;
                }

                if (lo == (size_t)n_tails) tails[n_tails++] = x;
                else tails[lo] = x;
            }

            checksum = (checksum * 31 + n_tails) % 1000000007;
        }
    }

    printf("arrays %d len %d passes %d checksum %ld\n", N_ARRAYS, LEN, PASSES, checksum);
    free(data);
    return 0;
}

File: leetcode/201-300/300-longest-increasing-subsequence/bench/lisscan.rs
// Benchmark mirror of lisscan.kara — LeetCode #300, Longest Increasing
// Subsequence. Same patience sorting, same hand-written binary search, same
// reused stack tails buffer. See ../README.md § Benchmarks.

const N_ARRAYS: usize = 3000;
const LEN: usize = 512;
const PASSES: usize = 24;
const SPREAD: i64 = 4096;

fn lcg(state: i64) -> i64 {
    (state.wrapping_mul(1103515245).wrapping_add(12345)) & 0x7fffffff
}

fn main() {
    let total = N_ARRAYS * LEN;
    let mut data = vec![0i64; total];

    let mut state: i64 = 20300;
    for i in 0..total {
        state = lcg(state);
        data[i] = (state / 65536) % SPREAD;
    }

    let mut tails = [0i64; LEN];
    let mut checksum: i64 = 0;

    for _pass in 0..PASSES {
        for a in 0..N_ARRAYS {
            let base = a * LEN;
            let mut n_tails: usize = 0;

            for k in 0..LEN {
                let x = data[base + k];

                let mut lo = 0usize;
                let mut hi = n_tails;
                while lo < hi {
                    let mid = lo + (hi - lo) / 2;
                    if tails[mid] < x { lo = mid + 1; } else { hi = mid; }
                }

                if lo == n_tails {
                    tails[n_tails] = x;
                    n_tails += 1;
                } else {
                    tails[lo] = x;
                }
            }

            checksum = (checksum * 31 + n_tails as i64) % 1_000_000_007;
        }
    }

    println!("arrays {} len {} passes {} checksum {}", N_ARRAYS, LEN, PASSES, checksum);
}
